### PR TITLE
Alerting: Added support for tuning rule evaluation using configuration options.

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
Add support for configuring 
- [execute_alerts](https://grafana.com/docs/grafana/latest/administration/configuration/#execute_alerts
)
- [max_attempts](https://grafana.com/docs/grafana/latest/administration/configuration/#max_attempts) and
- [min_interval_seconds](https://grafana.com/docs/grafana/latest/administration/configuration/#min_interval_seconds); the default value is `10` seconds (the scheduler's interval) while in the old alerting this was `1`.
- [evaluation_timeout_seconds](https://grafana.com/docs/grafana/latest/administration/configuration/#evaluation_timeout_seconds)

~I have introduced an additional option `copy_legacy_options`.~ if the above options have default values then they are ignored and the legacy options (under `alerting` section) are applied. This is so that we ensure same experience for migrated dashboard alerts.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: docs/sources/administration/configuration.md [hunk 1]
```diff
@@ -1128,11 +1128,11 @@ Sets the alert notification timeout. Default value is `30`.
 
 ### max_attempts
 
-Sets a maximum limit on attempts to sending alert notifications. Default value is `3`.
+Sets a maximum limit on attempts to sending alert notifications. This configuration is also supported in [Grafana 8 Alerts]({{< relref "../alerting/unified-alerting/_index.md" >}}. Default value is `3`.
 
 ### min_interval_seconds
 
-Sets the minimum interval between rule evaluations. Default value is `1`.
+Sets the minimum interval between rule evaluations. This configuration is also supported in Grafana 8 Alerts. For dashboard alerts the default value is `1`, for [Grafana 8 Alerts]({{< relref "../alerting/unified-alerting/_index.md" >}}#min-interval-seconds) the default value is `10` that equals each alerting system's scheduler interval.
```
Comment (kylebrandt) [inline docs/sources/administration/configuration.md:1135]: I think we should also specify that it must be a multiple of 10, else the default of 10 will be returned.

File: pkg/services/ngalert/ngalert.go [hunk 2]
```diff
@@ -132,3 +135,30 @@ func (ng *AlertNG) IsDisabled() bool {
...
+// getRuleMinIntervalSeconds returns the configured minimum rule interval.
+// If this value is less or equal to zero or not divided exactly by the scheduler interval
+// the scheduler interval (10 seconds) is returned.
+func (ng *AlertNG) getRuleMinIntervalSeconds() int64 {
+	if ng.Cfg.AlertingMinInterval <= 0 {
+		return baseIntervalSeconds // if it's not configured; apply default
+	}
+
+	if ng.Cfg.AlertingMinInterval%baseIntervalSeconds != 0 {
+		ng.Log.Debug("Configured minimum evaluation interval is not divided exactly by the scheduler interval and it will fallback to default", "alertingMinInterval", ng.Cfg.AlertingMinInterval, "baseIntervalSeconds", baseIntervalSeconds, "defaultIntervalSeconds", defaultIntervalSeconds)
```
Comment (kylebrandt) [inline pkg/services/ngalert/ngalert.go:159]: Should be Error level I think?

File: docs/sources/administration/configuration.md (hunk 1 above)
Comment (achatterjee-grafana) [reply docs/sources/administration/configuration.md:1135]: Plus one, as it was not clear to me.

**Review by achatterjee-grafana — APPROVED** [review]
Doc change looks ok, with one exception about the multiples.

File: pkg/setting/setting.go [hunk 3]
```diff
@@ -387,6 +387,10 @@ type Cfg struct {
 	// Live, -1 means unlimited connections.
 	LiveMaxConnections int
 
+	// Alertting
```
Comment (davidmparrott) [inline pkg/setting/setting.go:390]:
```suggestion
	// Alerting
```

**Review by davidmparrott — APPROVED** [review]

**Comment by gotjosh** [comment]
Sofia and I spoke a bit about this. We need to gather more feedback on whether we want to have all these options be able to change at runtime. In my eyes, only the evaluation timeout is worth it because depending on datasource latency you want to experiment with different values without restarting.

I think all the others can be moved to the new configuration section.

**Comment by papagian** [comment]
@gotjosh thinking a bit more, the difficulty if we have them in the new configuration section is how we guarantee the same experience for the migrated dashboard alerts?
Those dashboard alerts were tuned using the old configuration section (`alerting`) options and I don't see a way to migrate those settings under the new section.

**Comment by bergquist** [comment]
> We need to gather more feedback on whether we want to have all these options be able to change at runtime. 

If we make `max_attempts ` or `evaluation_timeout_seconds` available at runtime we need upper/lower bounds. 

We have cloud customers who asked us to change `min_interval_seconds` for them but very so not a strong reason to make it available at runtime IMO.

File: docs/sources/administration/configuration.md [hunk 4]
```diff
@@ -1115,6 +1115,24 @@ For more information about the Grafana 8 alerts, refer to [Unified Alerting]({{<
 
 Specify the frequency of polling for admin config changes. The default value is `60`.
 
+### execute_alerts
+
+Turns off alert rule execution, but Grafana 8 Alerts are still visible in the Grafana UI. Default is `true`. If it's true (the default) then the [respective legacy option]({{< relref "#execute_alerts-1">}}) is applied.
+
+### evaluation_timeout_seconds
+
+Sets the alert calculation timeout. Default value is `30`. If it's `30` (the default) then the [respective legacy option]({{< relref "#evaluation_timeout_seconds-1">}}) is applied.
```
Comment (kylebrandt) [inline docs/sources/administration/configuration.md:1124]: I forget if the dataproxy timeout applies here? If so may we should cross reference it since the queries will timeout even if this is higher (if that is true)

**Review by kylebrandt — APPROVED** [review]
Works as described for me :+1:

File: docs/sources/administration/configuration.md (hunk 4 above)
Comment (papagian) [reply docs/sources/administration/configuration.md:1124]: good point! I'll try to set a low proxy timeout to try it.

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/setting/setting_unified_alerting.go [hunk 5]
```diff
@@ -53,5 +53,33 @@ func (cfg *Cfg) ReadUnifiedAlertingSettings(iniFile *ini.File) error {
 		}
 	}
 
+	alerting := iniFile.Section("alerting")
```
Comment (gotjosh) [inline pkg/setting/setting_unified_alerting.go:56]: I think we can convey the meaning of what we're trying to do a bit better here.

```golang
// We want unified alerting settings to take precedence over the legacy ones. We check for defaults on the new setting to know whether we should override or not.
cfg.ReadLegacyAlertingSettingsWithWarning(iniFile)

unifiedAlertingExecuteAlerts := ua.Key("execute_alerts").MustBool(defaultAlertingExecuteAlerts)
if unifiedAlertingExecuteAlerts != defaultAlertingExecuteAlerts {
        cfg.UnifiedAlertingExecuteAlerts = unifiedAlertingExecuteAlerts
}

func (cfg *Cfg) ReadLegacyAlertingSettingsWithWarning(iniFile *ini.File) {
	alerting := iniFile.Section("alerting")
	cfg.UnifiedAlertingExecuteAlerts = alerting.Key("execute_alerts").MustBool(defaultAlertingExecuteAlerts)
	if cfg.UnifiedAlertingExecuteAlerts != defaultAlertingExecuteAlerts {
		cfg.Logger.Warn("using legacy setting of 'execute_alerts' please use the configuration option in the `unified_alerting` section.")
	}
	cfg.UnifiedAlertingEvaluationTimeout = time.Second * time.Duration(alerting.Key("evaluation_timeout_seconds").MustInt64(defaultAlertingEvaluationTimeoutSec))
	cfg.UnifiedAlertingMaxAttempts = alerting.Key("max_attempts").MustInt64(defaultAlertingMaxAttempts)
	cfg.UnifiedAlertingMinInterval = alerting.Key("min_interval_seconds").MustInt64(defaultUnifiedAlertingMinInterval)
}
```

The warning should be repeated for every setting we're trying to set.

File: pkg/services/ngalert/ngalert.go [hunk 6]
```diff
@@ -172,3 +175,30 @@ func (ng *AlertNG) IsDisabled() bool {
...
+	}
+	return defaultIntervalSeconds
+}
+
+// getRuleMinIntervalSeconds returns the configured minimum rule interval.
+// If this value is less or equal to zero or not divided exactly by the scheduler interval
+// the scheduler interval (10 seconds) is returned.
+func (ng *AlertNG) getRuleMinIntervalSeconds() int64 {
+	if ng.Cfg.UnifiedAlertingMinInterval <= 0 {
+		return defaultBaseIntervalSeconds // if it's not configured; apply default
```
Comment (gotjosh) [inline pkg/services/ngalert/ngalert.go:195]: This is already very hard to follow as is - do you think we can simplify this?

I'm thinking that both `ng.Cfg.UnifiedAlertingMinInterval <= 0` and `ng.Cfg.UnifiedAlertingMinInterval%defaultBaseIntervalSeconds` should be a hard error now at a configuration level, that way we can always make the assumption that the minimum internal is always going be >= 10 and will be divisible by the base default.

Technically, erroring out will be a breaking change at a configuration level but a very obvious one as opposed to causing surprises at runtime which sends us on a wild goose chase on issues while we try to reproduce. 

cc: @torkelo

**Review by gotjosh — CHANGES_REQUESTED** [review]
First pass, I have opinions on the wording about the documentation but I want to make sure we agree on the approach before we document it.

Ping me and I'll give you another review asap.

File: pkg/setting/setting.go [hunk 7]
```diff
@@ -419,14 +428,18 @@ type Cfg struct {
...
-	HAPushPullInterval             time.Duration
+	AdminConfigPollInterval          time.Duration
+	AlertmanagerConfigPollInterval   time.Duration
+	HAListenAddr                     string
+	HAAdvertiseAddr                  string
+	HAPeers                          []string
+	HAPeerTimeout                    time.Duration
+	HAGossipInterval                 time.Duration
+	HAPushPullInterval               time.Duration
+	UnifiedAlertingMaxAttempts       int64
```
[the remark this answers is not in the document]
Comment (papagian) [reply pkg/setting/setting.go:439]: Personally, I prefer be specific about it so that people don't get confused when browsing the code. I have noticed that people usually confuse the two systems eg when they report issues they don't specify which alerting system use. I feel that also other developers outside the alerting team also do.
In addition to this, given that we want to [move all settings to from package level vars in settings package to the setting.Cfg](https://github.com/grafana/grafana/tree/main/pkg#settings-refactoring), we have to move those of the legacy alerting as well.
However, I don't mind changing it, if we decide as a team that we prefer it.

File: pkg/setting/setting.go [hunk 8]
```diff
@@ -46,6 +46,15 @@ const (
 	ApplicationName  = "Grafana"
 )
 
+const (
+	defaultAdminConfigPollIntervalSec   int64 = 60
```
[the remark this answers is not in the document]
Comment (papagian) [reply pkg/setting/setting.go:50]: I will move it but do we have to export it?

File: pkg/setting/setting_test.go [hunk 9]
```diff
@@ -427,3 +428,125 @@ func TestGetCDNPathWithAlphaVersion(t *testing.T) {
...
+			verifyCfg: func(t *testing.T, cfg Cfg) {
+				require.Equal(t, time.Duration(defaultAdminConfigPollIntervalSec)*time.Second, cfg.AdminConfigPollInterval)
+				require.Equal(t, defaultAlertingMaxAttempts, cfg.UnifiedAlertingMaxAttempts)
+				require.Equal(t, defaultUnifiedAlertingMinInterval, cfg.UnifiedAlertingMinInterval)
+				require.Equal(t, defaultAlertingExecuteAlerts, cfg.UnifiedAlertingExecuteAlerts)
+				require.Equal(t, time.Duration(defaultAlertingEvaluationTimeoutSec)*time.Second, cfg.UnifiedAlertingEvaluationTimeout)
+			},
+		},
+		{
+			desc: "when unified alerting options are invalid, apply legacy options",
```
[the remark this answers is not in the document]
Comment (papagian) [reply pkg/setting/setting_test.go:508]: My impression it that we don't usually error if a configuration is invalid: we just apply a default. I believe that this is a deliberate choice so that the server does not fail to start in case of invalid configuration. @grafana/backend-platform WDYT?

File: pkg/setting/setting.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/setting/setting.go:50]: No, feel free to remove the export on the other's as well. I believe it was an oversight from my side.

File: pkg/setting/setting.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/setting/setting.go:439]: I have opened #39350 so that we can move to their own struct - WDYT?

File: pkg/setting/setting_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/setting/setting_test.go:508]: If that's the case then consider this resolved.

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/setting/setting_unified_alerting.go (hunk 5 above)
Comment (papagian) [reply pkg/setting/setting_unified_alerting.go:77]: @gotjosh
Reading this again I don't think that we want to warn users when the legacy option is the default (and made them to change it). In addition to this, these settings are loaded always (even if the unified alerting is disabled), this warning would be confusing in that case.
Instead, I think that we have to warn them when the new option (under unified_alerting) is default and we fallback in applying the legacy one.

File: conf/defaults.ini [hunk 10]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
 # The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 ha_push_pull_interval = 60s
 
+# Makes it possible to turn off alert rule execution but alerting UI is visible. If it's true (the default) then the respective legacy option is applied.
```
Comment (gotjosh) [inline conf/defaults.ini:767]:
```suggestion
# Enable or disable alerting rule execution. The alerting UI remains visible. This option has a legacy version in the `[alerting]` section that takes precedence.
```

File: conf/defaults.ini [hunk 11]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
...
+execute_alerts = true
+
+# Default setting for alert calculation timeout. Default value is 30.  If it's 30 (the default) then the respective legacy option is applied.
```
Comment (gotjosh) [inline conf/defaults.ini:770]:
```suggestion
# Alert evaluation timeout when fetching data from the datasource. It refers to the  This option has a legacy version in the `[alerting]` section that takes precedence.
```

File: pkg/setting/setting_unified_alerting.go [hunk 12]
```diff
@@ -50,34 +55,38 @@ type UnifiedAlertingSettings struct {
...
 	HAGossipInterval               time.Duration
 	HAPushPullInterval             time.Duration
+	MaxAttempts                    int64
+	MinInterval                    int64
+	EvaluationTimeout              time.Duration
+	ExecuteAlerts                  bool
 	DefaultConfiguration           string
 }
 
 func (cfg *Cfg) ReadUnifiedAlertingSettings(iniFile *ini.File) error {
```
Comment (gotjosh) [inline pkg/setting/setting_unified_alerting.go:65]:
```suggestion
// ReadUnifiedAlertingSettings reads both the `unified_alerting` and `alerting` sections of the configuration while preferring configuration the `alerting` section.
// It first reads the `unified_alerting` section, then looks for non-defaults on the `alerting` section and prefers those. 
func (cfg *Cfg) ReadUnifiedAlertingSettings(iniFile *ini.File) error {
```

File: conf/defaults.ini [hunk 13]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
...
+evaluation_timeout_seconds = 30
+
+# Default setting for max attempts to sending alert notifications. Default value is 3. If it's 3 (the default) then the respective legacy option is applied.
```
Comment (gotjosh) [inline conf/defaults.ini:773]:
```suggestion
# Number of times we'll attempt to evaluate an alert rule before giving up on that evaluation. This option has a legacy version in the `[alerting]` section that takes precedence.
```

File: conf/defaults.ini [hunk 14]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
 # The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 ha_push_pull_interval = 60s
 
+# Makes it possible to turn off alert rule execution but alerting UI is visible. If it's true (the default) then the respective legacy option is applied.
+execute_alerts = true
+
+# Default setting for alert calculation timeout. Default value is 30.  If it's 30 (the default) then the respective legacy option is applied.
+evaluation_timeout_seconds = 30
```
Comment (gotjosh) [inline conf/defaults.ini:771]: Can we please change this value to use the same nomenclature as the other intervals? That means:

1. Dropping the `seconds` suffix
2. Then, adding the time parsing.

File: conf/defaults.ini [hunk 15]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
...
+
+# Default setting for max attempts to sending alert notifications. Default value is 3. If it's 3 (the default) then the respective legacy option is applied.
+max_attempts = 3
+
+# Makes it possible to enforce a minimal interval between evaluations, to reduce load on the backend. Default value is 10 which is equal to the scheduler interval.  If it's 10 (the default) then the respective legacy option is applied.
```
Comment (gotjosh) [inline conf/defaults.ini:776]:
```suggestion
# Minimum interval to enforce between rule evaluations. Rules will be adjusted if they are less than this value. Higher values can help with resource management as we'll schedule fewer evaluations over time. This option has a legacy version in the `[alerting]` section that takes precedence.
```

File: conf/defaults.ini [hunk 16]
```diff
@@ -764,11 +764,24 @@ ha_gossip_interval = 200ms
...
+min_interval_seconds = 10
```
Comment (gotjosh) [inline conf/defaults.ini:777]: Can we please change this value to use the same nomenclature as the other intervals? That means:

Dropping the seconds suffix
Then, adding the time parsing.

File: docs/sources/administration/configuration.md [hunk 17]
```diff
@@ -1159,6 +1159,24 @@ across larger clusters at the expense of increased bandwidth usage. The default
 
 The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 
+### execute_alerts
```
Comment (gotjosh) [inline docs/sources/administration/configuration.md:1162]: Can we please match the format we have above?

Description, follow by the default in the form of `The default value is XX` and then for options that have a legacy option we can have something like:

- This option has a legacy version in the [respective legacy]({{< relref "#evaluation_timeout_seconds-1">}}) is section that takes precedence.

**Review by gotjosh — APPROVED** [review]
LGTM!

Approving to unblock. Feel free to choose which ones you'd like to do now and which ones you want to do on a follow-up PR. However, I do expect we tackle _all_ of them before 8.2.

File: pkg/setting/setting_unified_alerting.go (hunk 5 above)
Comment (papagian) [reply pkg/setting/setting_unified_alerting.go:105]: > if legacy is different from default and unified different from default - prefer legacy and print a warning

why in that case we have to prefer the legacy? if the unified is different from default means that deliberately somebody has changed it.

Comment (gotjosh) [reply pkg/setting/setting_unified_alerting.go:105]: My thinking here is that I want users to explicitly **remove** the legacy setting for the new setting to be active. That way we can guarantee the removal of the whole section and have their configuration not fail. 

Unless we don't error out when you have a setting set that doesn't exist?

File: docs/sources/administration/configuration.md [hunk 18]
```diff
@@ -1159,6 +1159,25 @@ across larger clusters at the expense of increased bandwidth usage. The default
...
+
+Enable or disable alerting rule execution. Default is `true`. The alerting UI remains visible. This option has a [legacy version in the alerting section]({{< relref "#execute_alerts-1">}}) that takes precedence.
+
+### evaluation_timeout
+
+Sets the alert evaluation timeout when fetching data from the datasource. Default value is `30s`. This option has a [legacy version in the alerting section]({{< relref "#evaluation_timeout_seconds">}}) that takes precedence.
+
```
Comment (gotjosh) [inline docs/sources/administration/configuration.md:1169]:
```suggestion

The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
```

File: pkg/setting/setting_unified_alerting.go (hunk 5 above)
Comment (papagian) [reply pkg/setting/setting_unified_alerting.go:105]: > Unless we don't error out when you have a setting set that doesn't exist?

we don't error in that case; we just never read it and use it.

File: docs/sources/administration/configuration.md (hunk 18 above)
Comment (gotjosh) [reply docs/sources/administration/configuration.md:1169]: This same comment applies to `min_interval`.

File: docs/sources/administration/configuration.md [hunk 19]
```diff
@@ -1159,6 +1159,25 @@ across larger clusters at the expense of increased bandwidth usage. The default
...
+### max_attempts
+
+Sets a maximum limit on attempts to sending alert notifications. Default value is `3`. If it's `3` (the default) then the [respective legacy option]({{< relref "#max_attempts-1">}}) is applied
+
+### min_interval
+
+Sets the minimum interval between rule evaluations. Default value is `10s` which equals the scheduler interval. This value should be multiple of `10`, otherwise it fallbacks to `10`. If it's `10` (the default) then the [respective legacy option]({{< relref "#min_interval_seconds-1">}}) is applied
```
Comment (gotjosh) [inline docs/sources/administration/configuration.md:1176]:
```suggestion

```

File: pkg/setting/setting_unified_alerting.go (hunk 5 above)
Comment (gotjosh) [reply pkg/setting/setting_unified_alerting.go:105]: TIL. I take my words back, let's go for preferring unified in that case then.

File: pkg/services/ngalert/ngalert.go (hunk 6 above)
Comment (papagian) [reply pkg/services/ngalert/ngalert.go:195]: https://github.com/grafana/grafana/discussions/39709

File: docs/sources/administration/configuration.md [hunk 20]
```diff
@@ -1159,6 +1159,28 @@ across larger clusters at the expense of increased bandwidth usage. The default
...
+
+The timeout string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
+
+### max_attempts
+
+Sets a maximum limit on attempts to sending alert notifications. Default value is `3`. If it's `3` (the default) then the [respective legacy option]({{< relref "#max_attempts-1">}}) is applied
+
+### min_interval
+
+Sets the minimum interval to enforce between rule evaluations. Default value is `10s` which equals the scheduler interval. Rules will be adjusted if they are less than this value or if they are not multiple of the scheduler interval (10s). Higher values can help with resource management as we'll schedule fewer evaluations over time. This option has [a legacy version in the alerting section]({{< relref "#min_interval_seconds">}}) that takes precedence.
```
Comment (gotjosh) [inline docs/sources/administration/configuration.md:1178]:
```suggestion
Sets the minimum interval to enforce between rule evaluations. The default value is `10s` which equals the scheduler interval. Rules will be adjusted if they are less than this value or if they are not multiple of the scheduler interval (10s). Higher values can help with resource management as we'll schedule fewer evaluations over time. This option has [a legacy version in the alerting section]({{< relref "#min_interval_seconds">}}) that takes precedence.
```

**Review by gotjosh — APPROVED** [review]
LGTM! A set of minor typos and wording for your next PR.
